pkg/operator/configobservation/etcdendpoints/observe_etcd_endpoints.go

```diff
@@ -0,0 +1,110 @@
+	return endpointsobserver.ObserveStorageURLs(genericListers, recorder, currentConfig)
+}
+
+// ObserveStorageURLs observes the storage config URLs. If there is a problem observing the current storage config URLs,
+// then the previously observed storage config URLs will be re-used.
+func ObserveStorageURLs(genericListers configobserver.Listers, recorder events.Recorder, currentConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	storageConfigURLsPath := []string{"storageConfig", "urls"}
+	var errs []error
+
+	previouslyObservedConfig := map[string]interface{}{}
+	currentEtcdURLs, found, err := unstructured.NestedStringSlice(currentConfig, storageConfigURLsPath...)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if found {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, currentEtcdURLs, storageConfigURLsPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	var etcdURLs []string
+	etcdEndpoints, err := listers.ConfigmapLister.ConfigMaps(etcdEndpointNamespace).Get(etcdEndpointName)
+	if errors.IsNotFound(err) {
+		recorder.Warningf("ObserveStorageFailed", "Required %s/%s configmap not found", etcdEndpointNamespace, etcdEndpointName)
+		return previouslyObservedConfig, append(errs, fmt.Errorf("configmaps/%s.%s: not found", etcdEndpointName, etcdEndpointNamespace))
+	}
+	if err != nil {
+		recorder.Warningf("ObserveStorageFailed", "Error getting %s/%s configmap: %v", etcdEndpointNamespace, etcdEndpointName, err)
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	// note: etcd bootstrap should never be added to the in-cluster kube-apiserver
+	// this can result in some early pods crashlooping, but ensures that we never contact the bootstrap machine from
+	// the in-cluster kube-apiserver so we can safely teardown out of order.
+
+	for k := range etcdEndpoints.Data {
+		address := etcdEndpoints.Data[k]
+		ip := net.ParseIP(address)
+		if ip == nil {
+			ipErr := fmt.Errorf("configmaps/%s in the %s namespace: %v is not a valid IP address", etcdEndpointName, etcdEndpointNamespace, address)
+			errs = append(errs, ipErr)
+			continue
+		}
+		// skip placeholder ip addresses used in previous versions where the hostname was used instead
+		if strings.HasPrefix(ip.String(), "192.0.2.") || strings.HasPrefix(ip.String(), "2001:db8:") {
+			// not considered an error
+			continue
+		}
+		// use the canonical representation of the ip address (not original input) when constructing the url
+		if ip.To4() != nil {
+			etcdURLs = append(etcdURLs, fmt.Sprintf("https://%s:2379", ip))
+		} else {
+			etcdURLs = append(etcdURLs, fmt.Sprintf("https://[%s]:2379", ip))
+		}
+	}
+
+	if len(etcdURLs) == 0 {
+		emptyURLErr := fmt.Errorf("configmaps %s/%s: no etcd endpoint addresses found", etcdEndpointNamespace, etcdEndpointName)
+		recorder.Warning("ObserveStorageFailed", emptyURLErr.Error())
+		errs = append(errs, emptyURLErr)
```

Comment:
this was probably cargo-culted, but in this condition, please return the previous config and errors here.

Reply:
Fixed

Reply:
This is actually codified in unit tests, which are now breaking

Reply:
I restored the original behavior because there are two explicit test cases around it and I'm not confident in changing it with this PR. If you and @hexfusion and @retroflexer are convinced the current behavior is wrong, let's describe the bug and do a followup.

Reply:
I actually decided to fix this after all, because it exposed a bug whereby the localhost was erroneously being appended regardless of whether endpoints were discovered.